air/src/constraints/chiplets/memory/mod.rs

```diff
@@ -123,47 +126,108 @@ fn enforce_d_inv<E: FieldElement>(
+    memory_flag_first_row: E,
 ) -> usize {
-    let mut index = 0;
-
-    // initialize memory to zero when reading from new context and address pair.
-    for i in 0..NUM_ELEMENTS {
-        result[index] = memory_flag * frame.init_read_flag() * frame.v(i);
-        index += 1;
-    }
-
-    // copy previous values when reading memory that was previously accessed.
-    for i in 0..NUM_ELEMENTS {
-        result[index] = memory_flag * frame.copy_read_flag() * (frame.v_next(i) - frame.v(i));
-        index += 1;
-    }
-
-    index
+    // c_i is set to 1 when `v'[i]` is not written to, and 0 otherwise.
+    //
+    // In other words, c_i is set to 1 when `v'[i]` needs to be constrained (to either 0 or `v[i]`).
+    //
+    // Note that `c_i` only uses values in the "next" row. This is because it must be used to
+    // constrain the first row of the memory chiplet, where that row sits in the "next" position of
+    // the frame, and the "current" row belongs to the previous chiplet (and hence the "current" row
+    // must not be accessed).
+    //
+    // As a result, `c_i` does not include the constraint of being in the memory chiplet, or in the
+    // same context and word - these must be enforced separately.
+    let (c0, c1, c2, c3) = {
+        // intuition: the i'th `f` flag is set to 1 when `i == 2 * idx1 + idx0`
+        let f0 = binary_not(frame.idx1_next()) * binary_not(frame.idx0_next());
+        let f1 = binary_not(frame.idx1_next()) * frame.idx0_next();
+        let f2 = frame.idx1_next() * binary_not(frame.idx0_next());
+        let f3 = frame.idx1_next() * frame.idx0_next();
+
+        let c_i = |f_i| {
+            // z_i is set to 1 when `v'[i]` is not being accessed.
+            let z_i = binary_not(frame.is_word_access_next()) * binary_not(f_i);
+            let is_read_next = frame.is_read_next();
+
+            is_read_next + binary_not(is_read_next) * z_i
+        };
+
+        (c_i(f0), c_i(f1), c_i(f2), c_i(f3))
+    };
+
+    // first row constraints: when row' is the first row, and v'[i] is not written to, then v'[i]
+    // must be 0.
+    result[0] = memory_flag_first_row * c0 * frame.v_next(0);
+    result[1] = memory_flag_first_row * c1 * frame.v_next(1);
+    result[2] = memory_flag_first_row * c2 * frame.v_next(2);
+    result[3] = memory_flag_first_row * c3 * frame.v_next(3);
+
+    // non-first row, new word or context constraints: when row' is a new word/ctx, and v'[i] is
+    // not written to, then v'[i] must be 0.
+    result[4] = memory_flag_no_last * binary_not(frame.f_scw_next()) * c0 * frame.v_next(0);
+    result[5] = memory_flag_no_last * binary_not(frame.f_scw_next()) * c1 * frame.v_next(1);
+    result[6] = memory_flag_no_last * binary_not(frame.f_scw_next()) * c2 * frame.v_next(2);
+    result[7] = memory_flag_no_last * binary_not(frame.f_scw_next()) * c3 * frame.v_next(3);
+
+    // non-first row, same word or context constraints: when row' is in the same word/ctx, and
```

Comment:
should it be `same word and context`?

Reply:
Yes, fixed. I also switched to "word address"